File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/Collections/CollList1.java
package Zeze.Transaction.Collections;

import java.util.Collection;
import Zeze.Transaction.Bean;
import Zeze.Transaction.Log;
import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.SerializeHelper;
import Zeze.Transaction.Record;
import Zeze.Transaction.Transaction;
import Zeze.Util.Reflect;
import org.pcollections.Empty;

public class CollList1<V> extends CollList<V> {
	protected final SerializeHelper.CodecFuncs<V> valueCodecFuncs;
	private final int logTypeId;

	public CollList1(Class<V> valueClass) {
		valueCodecFuncs = SerializeHelper.createCodec(valueClass);
		logTypeId = Zeze.Transaction.Bean.Hash32("Zeze.Transaction.LogList1<" + Reflect.GetStableName(valueClass) + '>');
	}

	private CollList1(int logTypeId, SerializeHelper.CodecFuncs<V> valueCodecFuncs) {
		this.valueCodecFuncs = valueCodecFuncs;
		this.logTypeId = logTypeId;
	}

	@Override
	public boolean add(V item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.Add(item);
		}
		var newList = _list.plus(item);
		if (newList == _list)
			return false;
		_list = newList;
		return true;
	}

	@Override
	public boolean remove(Object item) {
		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.Remove(item);
		}
		var newList = _list.minus(item);
		if (newList == _list)
			return false;
		_list = newList;
		return true;
	}

	@Override
	public void clear() {
		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			listLog.Clear();
		} else
			_list = org.pcollections.Empty.vector();
	}

	@Override
	public V set(int index, V item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.Set(index, item);
		}
		var old = _list.get(index);
		_list = _list.with(index, item);
		return old;
	}

	@Override
	public void add(int index, V item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			listLog.Add(index, item);
		} else
			_list = _list.plus(index, item);
	}

	@Override
	public V remove(int index) {
		if (isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.Remove(index);
		}
		var old = _list.get(index);
		_list = _list.minus(index);
		return old;
	}

	@Override
	public boolean addAll(Collection<? extends V> items) {
		// XXX
		for (var v : items) {
			if (null == v) {
				throw new NullPointerException();
			}
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.AddAll(items);
		}
		else {
			_list = _list.plusAll(items);
		}
		return true;
	}

	@Override
	public boolean removeAll(Collection<?> c) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			assert txn != null;
			txn.VerifyRecordAccessed(this);
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)txn.LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			return listLog.RemoveAll(c);
		}
		else {
			var oldList = _list;
			_list = _list.minusAll(c);
			return oldList != _list;
		}
	}

	@Override
	public LogBean CreateLogBean() {
		var log = new LogList1<>(logTypeId, valueCodecFuncs);
		log.setBelong(getParent());
		log.setThis(this);
		log.setVariableId(getVariableId());
		log.setValue(_list);
		return log;
	}

	@Override
	public void FollowerApply(Log _log) {
		@SuppressWarnings("unchecked")
		var log = (LogList1<V>)_log;
		for (var opLog : log.getOpLogs()) {
			switch (opLog.op) {
			case LogList1.OpLog.OP_MODIFY:
				_list = _list.with(opLog.index, opLog.value);
				break;
			case LogList1.OpLog.OP_ADD:
				_list = _list.plus(opLog.index, opLog.value);
				break;
			case LogList1.OpLog.OP_REMOVE:
				_list = _list.minus(opLog.index);
				break;
			case LogList1.OpLog.OP_CLEAR:
				_list = Empty.vector();
				break;
			}
		}
	}

	@Override
	protected void InitChildrenRootInfo(Record.RootInfo root) {
	}

	@Override
	public Bean CopyBean() {
		var copy = new CollList1<>(logTypeId, valueCodecFuncs);
		copy._list = _list;
		return copy;
	}

	@Override
	public void Encode(ByteBuffer bb) {
		var tmp = getList();
		bb.WriteUInt(tmp.size());
		var encoder = valueCodecFuncs.encoder;
		for (var e : tmp)
			encoder.accept(bb, e);
	}

	@Override
	public void Decode(ByteBuffer bb) {
		clear();
		var decoder = valueCodecFuncs.decoder;
		for (int i = bb.ReadUInt(); i > 0; i--)
			add(decoder.apply(bb));
	}
}

File: Zeze/Transaction/Collections/CollList1.cs

using System;
using System.Collections.Generic;

namespace Zeze.Transaction.Collections
{
	public class CollList1<E> : CollList<E>
	{
		public override E this[int index]
		{
			get => Data[index];
			set
			{
				if (value == null)
					throw new ArgumentNullException(nameof(value));

				if (this.IsManaged)
				{
					var txn = Transaction.Current;
					txn.VerifyRecordAccessed(this);
					var log = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
					log.SetItem(index, value);
				}
				else
				{
					_list = _list.SetItem(index, value);
				}
			}
		}

		public override void Add(E item)
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Add(item);
			}
			else
            {
				_list = _list.Add(item);
			}
		}

		public override void AddRange(IEnumerable<E> items)
		{
			// XXX
			foreach (var v in items)
			{
				if (null == v)
					throw new ArgumentNullException();
			}
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.AddRange(items);
			}
			else
            {
				_list = _list.AddRange(items);
			}
		}

		public override void Clear()
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Clear();
			}
			else
            {
				_list = System.Collections.Immutable.ImmutableList<E>.Empty;
			}
		}

		public override void Insert(int index, E item)
		{
			if (item == null)
				throw new ArgumentNullException();

			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Insert(index, item);
			}
			else
            {
				_list = _list.Insert(index, item);
			}
		}

		public override bool Remove(E item)
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				return listLog.Remove(item);
			}
			else
            {
				var tmp = _list.Remove(item);
				if (tmp == _list)
					return false;
				_list = tmp;
				return true;
			}
		}

		public override void RemoveAt(int index)
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.RemoveAt(index);
			}
			else
            {
				_list = _list.RemoveAt(index);
			}
		}

		public override void RemoveRange(int index, int count)
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var listLog = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.RemoveRange(index, count);
			}
			else
            {
				_list = _list.RemoveRange(index, count);
			}
		}

		public override LogBean CreateLogBean() {
            var log = new LogList1<E>
            {
                Belong = Parent,
                This = this,
                VariableId = VariableId,
                Value = _list
            };
            return log;
		}

		public override void FollowerApply(Log _log)
		{
			var log = (LogList1<E>)_log;
			foreach (var opLog in log.OpLogs)
			{
				switch (opLog.op)
				{
					case LogList1<E>.OpLog.OP_MODIFY:
						_list = _list.SetItem(opLog.index, opLog.value);
						break;
					case LogList1<E>.OpLog.OP_ADD:
						_list = _list.Insert(opLog.index, opLog.value);
						break;
					case LogList1<E>.OpLog.OP_REMOVE:
						_list = _list.RemoveAt(opLog.index);
						break;
					case LogList1<E>.OpLog.OP_CLEAR:
						_list = System.Collections.Immutable.ImmutableList<E>.Empty;
						break;
				}
			}
		}

		protected override void InitChildrenRootInfo(Record.RootInfo root)
		{
		}

		protected override void ResetChildrenRootInfo()
		{
		}

		public override CollList1<E> Copy()
		{
            var copy = new CollList1<E>
            {
                _list = _list
            };
            return copy;
		}
	}
}